### DISCUSSION TITLE
Any way in which we can increase the source document count?

### DISCUSSION THREAD

**Discussion by gaurav-cointab** [imartinez/privateGPT#416] [Q&A] [answered]
Any way in which we can increase the source document count?

Currently is says that it will only pick 4 documents snippet. Can we increase this to 10?

Should it be increased to 10? Is 4 enough.

**Comment by pseudotensor** [comment]
It's good point.  privateGPT already saturates the context with few-shot prompting from langchain.  In h2ogpt we optimized this more, and allow you to pass more documents if want via `k` CLI option.  Has UI and same CPU model support, as well as GPU support: https://github.com/h2oai/h2ogpt

**Comment by maozdemir** [comment]
> Any way in which we can increase the source document count?
> 
> Currently is says that it will only pick 4 documents snippet. Can we increase this to 10?
> 
> Should it be increased to 10? Is 4 enough.

There are no limitations in terms of the number of source documents in the current implementation. Please make sure that your files have the extensions in lowercase.

**Reply by pseudotensor** [reply]
He's talking about the number of chunks within the context window, not the total number of docs.  See my comment above.

**Comment by gaurav-cointab** [comment]
So all of you got the question wrong.

The current implementation goes something like this.

First our question is broken by embeddings. Then those values are searched in the chunked source documents. Then the filtered chunked source documents are sent to the llm along weight the query, and asked to prepare a result. In response we are getting the answer, and along with that we are getting the 4 chunked source documents which were sent to llm for answer preparing.

Now if in response we are getting 4 chunked source documents, then those are the 4 which were picked by embedding to be sent to llm. I want to increase this number.

Is this doable?

**Reply by pseudotensor** [reply, marked as answer]
That's what I was saying.  In h2oGPT one has just pass `k` as a parameter like `python generate.py -k=10` and it'll give 10 document chunks to LLM.

You can google and see how to do k for privateGPT.  It's like:
```
retriever = db.as_retriever(search_type="similarity", search_kwargs={"k":10})
```

**Reply by pseudotensor** [reply]
Example google search: https://towardsdatascience.com/4-ways-of-question-answering-in-langchain-188c6707cc5a

**Reply by pseudotensor** [reply]
With more chunks you will have to make each chunk smaller to fit into context, however.

**Reply by gaurav-cointab** [reply]
This is awesome. Thanks so much. Sorry for not acknowledging you answer earlier. I just few minutes read it back again, and found that you were saying what I wanted, but since it had some technical terms, it might not struck me at that point in time.

But these links and articles, are awesome, and have given me a lot of try and experiment on.
